modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+        client.admin().indices().delete(deleteIndexRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                listener.onResponse(null);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                // we're capturin
+                logger.error(() -> format("DLM request to delete [%s] failed", targetIndex), e);
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    @Nullable
+    static TimeValue getRetentionConfiguration(DataStream dataStream) {
+        TimeValue retention = null;
+        if (dataStream.getLifecycle() != null) {
+            retention = dataStream.getLifecycle().getDataRetention();
+        }
+        return retention;
+    }
+
+    /**
+     * Calculate the age of the index since creation or rollover time if the index was already rolled.
+     * The rollover target is the data stream name the index is a part of.
+     */
+    static TimeValue getCreationOrRolloverDate(String rolloverTarget, IndexMetadata index) {
+        RolloverInfo rolloverInfo = index.getRolloverInfos().get(rolloverTarget);
+        if (rolloverInfo != null) {
+            return TimeValue.timeValueMillis(rolloverInfo.getTime());
+        } else {
+            return TimeValue.timeValueMillis(index.getCreationDate());
+        }
+    }
+
+    /**
+     * This is quite a shallow method but the purpose of its existence is to have only one place to modify once we
+     * introduce the index.lifecycle.prefer_ilm setting. Once the prefer_ilm setting exists the method will also
+     * make more sense as it will encapsulate a bit more logic.
+     */
+    private static boolean isManagedByDLM(DataStream parentDataStream, IndexMetadata indexMetadata) {
+        return indexMetadata.getLifecyclePolicyName() == null && parentDataStream.getLifecycle() != null;
+    }
+
+    private RolloverRequest getDefaultRolloverRequest(String dataStream) {
+        RolloverRequest rolloverRequest = new RolloverRequest(dataStream, null).masterNodeTimeout(TimeValue.MAX_VALUE);
+
+        // TODO get rollover from cluster setting once we have it
+        rolloverRequest.addMaxIndexAgeCondition(TimeValue.timeValueDays(30));
+        rolloverRequest.addMaxPrimaryShardSizeCondition(ByteSizeValue.ofGb(50));
+        rolloverRequest.addMaxPrimaryShardDocsCondition(200_000_000);
+        // don't rollover an empty index
+        rolloverRequest.addMinIndexDocsCondition(1);
+        return rolloverRequest;
+    }
+
+    private void onBecomeMaster() {
+        maybeScheduleJob();
+    }
+
+    private void updatePollInterval(TimeValue newInterval) {
+        this.pollInterval = newInterval;
+        maybeScheduleJob();
+    }
+
+    private void cancelJob() {
+        if (scheduler.get() != null) {
+            scheduler.get().remove(DATA_LIFECYCLE_JOB_NAME);
```

Comment:
One of these days we should just write a `SetOnceOptional<T>` class of our own that combines `SetOnce` and `Optional`, so we can dispense with the safety checks and do `scheduler.set(...)` and `scheduler.ifPresent(s -> s.remove(DATA_LIFECYCLE_JOB_NAME));`